### PR TITLE
JAMES-2881 Add test for HeadersToHTTP/SerialiseToHTTP mailets

### PR DESCRIPTION
**Pull request by splainez**
Added tests, and fixed errors in the request of SerialiseToHTTP

https://issues.apache.org/jira/browse/JAMES-2881

### CODE REVIEW

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/SerialiseToHTTP.java [hunk 1]
```diff
@@ -118,16 +118,14 @@ public void init() throws MessagingException {
...
+            String result = httpPost(
+                    getNameValuePairs(MimeMessageUtil.asString(mail.getMessage())));
             if (passThrough) {
                 addHeader(mail, (result == null || result.length() == 0), result);
             } else {
                 mail.setState(Mail.GHOST);
             }
-        } catch (MessagingException | IOException me) {
+        } catch (Exception me) {
             LOGGER.error("Messaging exception", me);
```
Comment (chibenwa) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/SerialiseToHTTP.java]: It is no longer a Messaging exception.

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/HeadersToHTTPTest.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
+
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/HeadersToHTTPTest.java]: You are missing the Apache2 compulsory license.

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/HeadersToHTTPTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+public class HeadersToHTTPTest {
+
+    private static HttpServer server;
+    private static UriHttpRequestHandlerMapper mapper;
+    private Mail mail;
+
+    private String urlTestPattern;
+
+    @BeforeAll
+    public static void setupServer() throws MessagingException, IOException {
```
Comment (chibenwa) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/HeadersToHTTPTest.java]: With junit 5 `public` qualifiers can safely be omited.

File: mailet/standard/pom.xml [hunk 4]
```diff
@@ -142,6 +142,10 @@
             <groupId>org.slf4j</groupId>
             <artifactId>slf4j-api</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>jcl-over-slf4j</artifactId>
```
Comment (mbaechler) [inline mailet/standard/pom.xml]: we try to keep our dependencies ordered (first by groupId then by artifactId). However we split dependencies in two categories: internal then externals.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/SerialiseToHTTP.java [hunk 5]
```diff
@@ -118,16 +118,14 @@ public void init() throws MessagingException {
...
-            String result = httpPost(nameValuePairs);
+            String result = httpPost(
+                    getNameValuePairs(MimeMessageUtil.asString(mail.getMessage())));
             if (passThrough) {
                 addHeader(mail, (result == null || result.length() == 0), result);
             } else {
                 mail.setState(Mail.GHOST);
             }
-        } catch (MessagingException | IOException me) {
+        } catch (Exception me) {
```
Comment (mbaechler) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/SerialiseToHTTP.java]: This exception is too broad in my opinion

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/HeadersToHTTPTest.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        mapper.register(urlTestPattern, (request, response, context) -> {
+
+            assertThat(request.getRequestLine().getMethod()).isEqualTo("POST");
+
+            BasicHttpEntityEnclosingRequest basicRequest = (BasicHttpEntityEnclosingRequest) request;
+            BasicHttpEntity entity = (BasicHttpEntity) basicRequest.getEntity();
+
+            try {
+                List<NameValuePair> params = URLEncodedUtils.parse(entity);
+                assertThat(params).hasSize(5).anySatisfy((param) -> {
```
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/HeadersToHTTPTest.java]: these assertions belong to another test in my opinion (it doesn't match "shoud succeed" description)

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/SerialiseToHTTPTest.java [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+                .build();
+
+        mapper.register(urlTestPattern, (request, response, context) -> {
+            assertThat(request.getRequestLine().getMethod()).isEqualTo("POST");
+
+            BasicHttpEntityEnclosingRequest basicRequest = (BasicHttpEntityEnclosingRequest) request;
+            BasicHttpEntity entity = (BasicHttpEntity) basicRequest.getEntity();
+
+            try {
+                List<NameValuePair> params = URLEncodedUtils.parse(entity);
```
Comment (mbaechler) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/SerialiseToHTTPTest.java]: same comment here

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/SerialiseToHTTP.java (hunk 5 above)
Comment (splainez) [reply mailet/standard/src/main/java/org/apache/james/transport/mailets/SerialiseToHTTP.java]: This code snippet it's outdate, but i agree with you, in the new code, i try to isolate the exception only for the call to `MimeMessageUtil.asString(message);` this method throws Exception, however, if you think it's not correct, i can write a new method in the class, and only throw Messaging and IO exceptions.

File: mailet/standard/src/main/java/org/apache/james/transport/mailets/SerialiseToHTTP.java [hunk 8]
```diff
@@ -117,19 +117,28 @@ public void init() throws MessagingException {
...
+            messageAsString = MimeMessageUtil.asString(message);
+        } catch (Exception e) {
+            LOGGER.error("Message to string exception", e);
+            addHeader(mail, false, e.getMessage());
+            return;
+        }
+
+        String result = httpPost(getNameValuePairs(messageAsString));
+        if (passThrough) {
+            addHeader(mail, (result == null || result.length() == 0), result);
```
Comment (blackheaven) [inline mailet/standard/src/main/java/org/apache/james/transport/mailets/SerialiseToHTTP.java]: `StringUtils.isEmpty`

File: mailet/standard/src/test/java/org/apache/james/transport/mailets/HeadersToHTTPTest.java [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        FakeMailetConfig mailetConfig = FakeMailetConfig.builder()
+                .setProperty("parameterKey", "pKey").setProperty("parameterValue", "pValue")
+                .setProperty("url", "http://qwerty.localhost:123456" + urlTestPattern).build();
+
+        Mailet mailet = new HeadersToHTTP();
+        mailet.init(mailetConfig);
+
+        mailet.service(mail);
+
+        assertThat(mail.getMessage().getHeader("X-headerToHTTP")).hasSize(1)
```
Comment (blackheaven) [inline mailet/standard/src/test/java/org/apache/james/transport/mailets/HeadersToHTTPTest.java]: `hasSize` on a dedicated line

**Comment by chibenwa** [comment]
@blackheaven I solved your issues on a separated pull request: https://github.com/linagora/james-project/pull/2709

**Comment by chibenwa** [comment]
Hi,

This pull request has just been merged.

Thank you again for your dedication toward the James project.

Cheers,

Benoit

**Pull request closed without merging**